src/worker_thread.c

```diff
@@ -25,42 +25,48 @@ typedef struct WORKER_THREAD_TAG
     WORKER_FUNC worker_func;
     void* worker_func_context;
     THREAD_HANDLE thread_handle;
-    HANDLE worker_thread_shutdown_event;
-    HANDLE execute_worker_event;
+    volatile_atomic int32_t state;
+
     SM_HANDLE sm;
 } WORKER_THREAD;
 
+#define WORKER_THREAD_STATE_VALUES \
+    WORKER_THREAD_STATE_IDLE, \
+    WORKER_THREAD_STATE_PROCESS_ITEM, \
+    WORKER_THREAD_STATE_CLOSE
+
+MU_DEFINE_ENUM(WORKER_THREAD_STATE, WORKER_THREAD_STATE_VALUES)
+
 static int worker_thread_func(void* arg)
 {
     WORKER_THREAD_HANDLE worker_thread = arg;
     int result = 0;
-    HANDLE wait_events[2];
-
-    wait_events[0] = worker_thread->worker_thread_shutdown_event;
-    wait_events[1] = worker_thread->execute_worker_event;
+    bool continue_run = true;
 
-    while (1)
+    do
     {
-        /* Codes_SRS_WORKER_THREAD_01_019: [ The worker thread started by worker_thread_create shall wait for the 2 events: thread shutdown event and execute worker function event. ]*/
-        DWORD wait_result = WaitForMultipleObjects(2, wait_events, FALSE, INFINITE);
-        if (wait_result == WAIT_OBJECT_0)
+        // Codes_SRS_WORKER_THREAD_01_019: [ The worker thread started by worker_thread_create shall get the module state. ]
+        WORKER_THREAD_STATE current_state = interlocked_add(&worker_thread->state, 0);
+        switch (current_state)
         {
-            /* Codes_SRS_WORKER_THREAD_01_020: [ When the shutdown event is signaled, the worker thread shall exit. ]*/
-            result = 0;
-            break;
+            default:
+            case WORKER_THREAD_STATE_IDLE:
+                // Codes_SRS_WORKER_THREAD_11_002: [ If the module state is WORKER_THREAD_STATE_IDLE, the worker thread shall wait for the state to transition to something else. ]
+                (void)wait_on_address(&worker_thread->state, current_state, UINT32_MAX);
+                break;
+            case WORKER_THREAD_STATE_PROCESS_ITEM:
+                // Codes_SRS_WORKER_THREAD_11_001: [ ... and set the module state to idle if it has not been changed. ]
+                (void)interlocked_compare_exchange(&worker_thread->state, WORKER_THREAD_STATE_IDLE, current_state);
```

Comment:
>interlocked_compare_exchange [](http://example.com/codeflow?start=22&length=28)

why not interlocked_exchange? it doesn't look like the "current_state" is examined.

Reply:
I do not want to change the value if it has been changed somewhere else.  For example between line 49 and 59 the user calls close, I don't want to change it back to IDLE, I want this call to not change the value and then handle the close call in the next iteration of the loop.